evals/api.py

```diff
@@ -23,85 +24,139 @@
 logger = logging.getLogger(__name__)
 
 
-def completion_query(
-    model_spec: ModelSpec,
-    prompt: Union[OpenAICreatePrompt, OpenAICreateChatPrompt, Prompt],
-    **kwargs,
-) -> tuple[dict, Union[OpenAICreatePrompt, OpenAICreateChatPrompt], dict]:
-    """
-    Query the API for a completion.
+class ReturnType(ABC):
+    def __init__(self, raw_data: Any, prompt: Any, metadata: Dict[str, Any] = None):
+        self.raw_data = raw_data
+        self.prompt: str = prompt
+        self.completions: List[str] = self.extract_completions()
+        self.metadata = metadata if metadata is not None else {}
 
-    ARGS
-    ====
-    `model_spec`: `ModelSpec` containing model details to use in the query.
-        This should be the dict returned by `registry.get_model()`.
-        If `model_spec` is not provided, we use the default model that was
-            intialized at the beginning of the run.
-    `prompt`: Either a `Prompt` object or a raw prompt that will get wrapped in
-        the approriate `Prompt` class.
-    `kwargs`: Other arguments passed to the API.
+    @abstractmethod
+    def extract_completions(self) -> List[str]:
+        pass
 
-    RETURNS
-    =======
-    The result of the API call.
-    The prompt that was fed into the API call as a str.
-    A dict containing metadata about the query.
-    """
-    if not isinstance(prompt, Prompt):
-        assert (
-            isinstance(prompt, str)
-            or (isinstance(prompt, list) and all(isinstance(token, int) for token in prompt))
-            or (isinstance(prompt, list) and all(isinstance(token, str) for token in prompt))
-            or (isinstance(prompt, list) and all(isinstance(msg, dict) for msg in prompt))
-        ), f"Got type {type(prompt)}, with val {type(prompt[0])} for prompt, expected str or list[int] or list[str] or list[dict[str, str]]"
+
+class CompletionFn(Protocol):
```

Comment:
The name CompletionFn feels like a clunky, do you have any better ideas? Though it does make sense as it currently stands.. it's a function that does a completion.

Also, I'm wondering if we want to have this abstraction support other ways of interacting with models like pick-style evals, where given a set of options (e.g. A/B/C) the model selects the most likely option using logprobs. Context is that GPT-3 (and smaller) has a hard time sampling A/B/C, so we used pick evals in our GPT-3 paper.

Reply:
If we want to support GPT-3 or similar (possibly external) models, we would probably need to support pick evals. So I'd say that the question is more like.. What kind of models we'd like to support?

Reply:
How about `CompletionProtocol` or `PromptCompletionProtocol`?

Reply:
Hmm I still prefer `CompletionFn` over those two options. Maybe we can leave as is for now?